File: leetcode/LRUCache.java
class LRUCache {
	private int capacity;
	private HashMap cacheMap = new HashMap();
	private Item head;
	private Item tail;

    public LRUCache(int capacity) {
		this.capacity = capacity;
    }
    
    public int get(int key) {
		if (!cacheMap.containsKey(key))
			return -1;
		touch(key);
		return ((Item)cacheMap.get(key)).value;
    }
    
    public void put(int key, int value) {
		if (cacheMap.containsKey(key)) {
			((Item)cacheMap.get(key)).value = value;
			touch(key);
			return;
		}
		removeLeaseRecent();
		putNew(new Item(key, value));
    }
	private void touch(int key) {
		if (cacheMap.size() != 1 && key != tail.key) {
			Item latest = (Item)cacheMap.get(key);

			if (key == head.key) {
				head = latest.next;
				tail = latest;

			} else {
				// disconnect
				latest.prev.next = latest.next;
				latest.next.prev = latest.prev;

				// append to tail
				latest.prev = tail;
				latest.next = head;

				// update head/tail
				tail.next = latest;
				tail = latest;
				head.prev = latest;
			}
		}
	}

	private void removeLeaseRecent() {
		if (cacheMap.size() == capacity) {
			if (capacity == 1) {
				cacheMap.remove(head.key);
				return;
			}
			Object k = head.key;
			head.next.prev = tail;
			tail.next = head.next;
			head = head.next;
			cacheMap.remove(k);
		}
	}

	private void putNew(Item item) {
		if (cacheMap.size() == 0) {
			head = item;
			tail = item;
		} else {
			item.prev = tail;
			item.next = head;

			tail.next = item;
			head.prev = item;

			tail = item;

			if (head.next == null)
				head.next = item;
		}

		cacheMap.put(item.key, item);
	}
    private class Item {
		public int key;
		public int value;
		public Item prev;
		public Item next;

		public Item(int key, int value) {
			this.key = key;
			this.value = value;
		}
	}

}

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache obj = new LRUCache(capacity);
 * int param_1 = obj.get(key);
 * obj.put(key,value);
 */

File: leetcode/LRUCache.py
class Item(object):
    def __init__(self, key, value):
        self.key = key
        self.value = value
        
        self.prev = None
        self.next = None

class LRUCache(object):

    def __init__(self, capacity):
        """
        :type capacity: int
        """
        self.capacity = capacity
        self.cacheMap = {}
        self.head = None
        self.tail = None
        

    def get(self, key):
        """
        :type key: int
        :rtype: int
        """
        if key not in self.cacheMap:
            return -1
        self.touch(key)
        return self.cacheMap[key].value
    
    def touch(self, key):
        if self.size() != 1 and key != self.tail.key:
            latest = self.cacheMap[key]
            
            if key == self.head.key:
                self.head = latest.next
                self.tail = latest
                
            else:
                
                # disconnect
                latest.prev.next = latest.next
                latest.next.prev = latest.prev
                
                # append to tail
                latest.prev = self.tail
                latest.next = self.head
                
                # update head/tail
                self.tail.next = latest
                self.tail = latest
                self.head.prev = latest

    def put(self, key, value):
        """
        :type key: int
        :type value: int
        :rtype: None
        """
        if key in self.cacheMap:
            self.cacheMap[key].value = value
            self.touch(key)
            return

        self.remove_least()
        self.put_new(Item(key, value))
            
    def remove_least(self):
        if self.size() == self.capacity:
            if self.capacity == 1:
                del self.cacheMap[self.head.key]
                return
            k = self.head.key
            self.head.next.prev = self.tail
            self.tail.next = self.head.next
            self.head = self.head.next
            del self.cacheMap[k]
    
    def put_new(self, item):
        if self.size() == 0:
            self.head = item
            self.tail = item
        else:
            item.prev = self.tail
            item.next = self.head
            
            self.tail.next = item
            self.head.prev = item
            
            self.tail = item

            if self.head.next is None:
                self.head.next = item

        self.cacheMap[item.key] = item
        
    def size(self):
        return len(self.cacheMap)

